lib/mysqlDao.js

```diff
@@ -193,31 +208,76 @@ class MysqlDao {
 
     /**
      * @param {object} whereClause
+     * @param {MysqlDaoGetAllOptions}
      * @returns {Promise.<object[],Error>}
-     * 
-     * @TODO: add an options field that allows order by, limit, arbitrary WHERE clause to AND / OR together
      */
-    getAll(whereClause) {
-        return this._getAll(whereClause);
+    getAll(whereClause, options) {
+        options = options || {};
+        return this._getAll(whereClause, options);
     }
 
     /**
      * @param {object} whereClause
+     * @param {MysqlDaoGetAllOptions}
+     * @returns {Promise.<MysqlDaoGetAllAndCountResult,Error>}
+     */
+    getAllAndCount(whereClause, options) {
+        options = options || {};
+        return this._getAllAndCount(whereClause, options);
+    }
+
+    /**
+     * @param {object} whereClause
+     * @param {MysqlDaoGetAllOptions}
      * @returns {Promise.<object[],Error>}
      */
-    getAllFromMaster(whereClause) {
-        return this._getAll(whereClause, {forceReadFromMaster: true});
+    getAllFromMaster(whereClause, options) {
+        options = options || {};
+        return this._getAll(whereClause, options, {forceReadFromMaster: true});
+    }
+
+    /**
+     * @param {object} whereClause
+     * @param {MysqlDaoGetAllOptions}
+     * @returns {Promise.<MysqlDaoGetAllAndCountResult,Error>}
+     */
+    getAllAndCountFromMaster(whereClause, options) {
+        options = options || {};
+        return this._getAllAndCount(whereClause, options, {forceReadFromMaster: true});
     }
 
     /**
      * @param whereClause
-     * @param {MysqlGetFromSqlOptions} [options]
+     * @param {MysqlDaoGetAllOptions}
+     * @param {MysqlGetFromSqlOptions} [config]
      * @returns {Promise.<Object[], Error>}
      * @private
      */
-    _getAll(whereClause, options) {
-        const sql = this._queryHelper.getAll(whereClause);
-        return this._getAllFromSql(sql, options);
+    _getAll(whereClause, options, config) {
+        if (options && options.offset && !options.limit) throw new SyntaxError('Missing limit in getAll options, must include with offset option.');
+        const sql = this._queryHelper.getAll(whereClause, options);
+        return this._getAllFromSql(sql, config);
+    }
+
+    /**
+     * @param whereClause
+     * @param {MysqlDaoGetAllOptions}
+     * @param {MysqlGetFromSqlOptions} [config]
+     * @returns {Promise.<MysqlDaoGetAllAndCountResult, Error>}
+     * @private
+     */
+    _getAllAndCount(whereClause, options, config) {
+        config = config || {};
+        if (options && options.offset && !options.limit) throw new SyntaxError('Missing limit in getAll options, must include with offset option.');
```

Comment:
I don't think this is a syntax error.  https://nodejs.org/api/errors.html#errors_class_syntaxerror  You can just use the Error class.